Source/DataSources/GpxDataSource.js

```diff
@@ -0,0 +1,1022 @@
+import Cartesian2 from "../Core/Cartesian2.js";
+import Cartesian3 from "../Core/Cartesian3.js";
+import ClockRange from "../Core/ClockRange.js";
+import ClockStep from "../Core/ClockStep.js";
+import Color from "../Core/Color.js";
+import createGuid from "../Core/createGuid.js";
+import defaultValue from "../Core/defaultValue.js";
+import defined from "../Core/defined.js";
+import DeveloperError from "../Core/DeveloperError.js";
+import Event from "../Core/Event.js";
+import Iso8601 from "../Core/Iso8601.js";
+import JulianDate from "../Core/JulianDate.js";
+import NearFarScalar from "../Core/NearFarScalar.js";
+import PinBuilder from "../Core/PinBuilder.js";
+import Resource from "../Core/Resource.js";
+import RuntimeError from "../Core/RuntimeError.js";
+import TimeInterval from "../Core/TimeInterval.js";
+import TimeIntervalCollection from "../Core/TimeIntervalCollection.js";
+import HeightReference from "../Scene/HeightReference.js";
+import HorizontalOrigin from "../Scene/HorizontalOrigin.js";
+import LabelStyle from "../Scene/LabelStyle.js";
+import VerticalOrigin from "../Scene/VerticalOrigin.js";
+import Autolinker from "../ThirdParty/Autolinker.js";
+import when from "../ThirdParty/when.js";
+import BillboardGraphics from "./BillboardGraphics.js";
+import ConstantProperty from "./ConstantProperty.js";
+import DataSource from "./DataSource.js";
+import DataSourceClock from "./DataSourceClock.js";
+import EntityCluster from "./EntityCluster.js";
+import EntityCollection from "./EntityCollection.js";
+import LabelGraphics from "./LabelGraphics.js";
+import PolylineGraphics from "./PolylineGraphics.js";
+import PolylineOutlineMaterialProperty from "./PolylineOutlineMaterialProperty.js";
+import SampledPositionProperty from "./SampledPositionProperty.js";
+
+let parser;
+if (typeof DOMParser !== "undefined") {
+  parser = new DOMParser();
+}
+
+const autolinker = new Autolinker({
+  stripPrefix: false,
+  email: false,
+  replaceFn: function (linker, match) {
+    if (!match.protocolUrlMatch) {
+      // Prevent matching of non-explicit urls.
+      // i.e. foo.id won't match but http://foo.id will
+      return false;
+    }
+  },
+});
+
+const BILLBOARD_SIZE = 32;
+const BILLBOARD_NEAR_DISTANCE = 2414016;
+const BILLBOARD_NEAR_RATIO = 1.0;
+const BILLBOARD_FAR_DISTANCE = 1.6093e7;
+const BILLBOARD_FAR_RATIO = 0.1;
+
+const gpxNamespaces = [null, undefined, "http://www.topografix.com/GPX/1/1"];
+const namespaces = {
+  gpx: gpxNamespaces,
+};
+
+function readBlobAsText(blob) {
+  const deferred = when.defer();
+  const reader = new FileReader();
+  reader.addEventListener("load", function () {
+    deferred.resolve(reader.result);
+  });
+  reader.addEventListener("error", function () {
+    deferred.reject(reader.error);
+  });
+  reader.readAsText(blob);
+  return deferred;
```

Comment:
`KmlDataSource` returns `deferred.promise` in a similar function. While the deferred object works in most cases I believe they don't have all of the same methods as `always`.